libvips/foreign/openslideconnection.c

```diff
@@ -0,0 +1,252 @@
+}
+
+static void
+vips_openslideconnection_oldest_cb(gpointer key, gpointer value, gpointer data)
+{
+	VipsOpenslideConnection *connection = (VipsOpenslideConnection *) value;
+	VipsOpenslideConnection **oldest = (VipsOpenslideConnection **) data;
+
+	g_mutex_lock(&connection->lock);
+
+	if (!*oldest ||
+		connection->time < (*oldest)->time)
+		*oldest = connection;
+
+	g_mutex_unlock(&connection->lock);
+}
+
+static VipsOpenslideConnection *
+vips_openslideconnection_oldest(void)
+{
+	VipsOpenslideConnection *oldest;
+
+	oldest = NULL;
+	g_hash_table_foreach(vips_openslideconnection_cache,
+		vips_openslideconnection_oldest_cb, &oldest);
+
+	return oldest;
+}
+
+static void
+vips_openslideconnection_trim(void)
+{
+	VipsOpenslideConnection *oldest;
+
+	while (vips_openslideconnection_unused() > OPENSLIDECONNECTION &&
+		(oldest = vips_openslideconnection_oldest()))
+		vips_openslideconnection_free(oldest);
+}
+
+static void
+vips_openslideconnection_unref(VipsOpenslideConnection *connection)
+{
+	gboolean trim;
+
+	g_mutex_lock(&connection->lock);
+
+	g_assert(connection->ref_count > 0);
+
+	connection->ref_count -= 1;
+	trim = connection->ref_count == 0;
+
+	g_mutex_unlock(&connection->lock);
+
+	if (trim)
+		vips_openslideconnection_trim();
+}
+
+static void
+vips_openslideconnection_ref(VipsOpenslideConnection *connection)
+{
+	g_mutex_lock(&connection->lock);
+
+	g_assert(connection->ref_count >= 0);
+
+	connection->ref_count += 1;
+
+	g_mutex_unlock(&connection->lock);
+}
+
+static VipsOpenslideConnection *
+vips_openslideconnection_new(const char *filename)
+{
+	VipsOpenslideConnection *connection;
+
+	connection = g_new0(VipsOpenslideConnection, 1);
+	connection->filename = g_strdup(filename);
+
+	g_assert(!g_hash_table_lookup(vips_openslideconnection_cache, filename));
+
+	g_hash_table_insert(vips_openslideconnection_cache,
+		connection->filename, connection);
+
+	return connection;
+}
+
+static void
+vips_openslideconnection_touch(VipsOpenslideConnection *connection)
+{
+	static int time = 0;
+
+	g_mutex_lock(&connection->lock);
+
+	connection->time = time++;
+
+	g_mutex_unlock(&connection->lock);
+}
+
+openslide_t *
+vips__openslideconnection_open(const char *filename)
+{
+	g_mutex_lock(&vips_openslideconnection_lock);
+
+	if (!vips_openslideconnection_cache)
+		vips_openslideconnection_cache =
+			g_hash_table_new(g_str_hash, g_str_equal);
+
+	VipsOpenslideConnection *connection;
+
+	connection = g_hash_table_lookup(vips_openslideconnection_cache, filename);
+
+	if (!connection)
+		connection = vips_openslideconnection_new(filename);
+
+	vips_openslideconnection_ref(connection);
+	vips_openslideconnection_touch(connection);
+
+	g_mutex_unlock(&vips_openslideconnection_lock);
+
+	g_mutex_lock(&connection->lock);
+
+	/* We do the open outside the main lock (just in the connection lock)
+	 * since it can take many seconds for some slides.
+	 */
+	if (!connection->osr)
+		connection->osr = openslide_open(connection->filename);
```

Comment:
`openslide_open` can validly return NULL, in which case this will rerun on every open attempt.  Probably fine, since the NULL case is an early exit which is pretty cheap (equivalent to `openslide_detect_vendor`).

Reply:
Ah it'll ref again on the second failure. I'll fix this too.